docs/spec/type_system.rst

```diff
@@ -4,5 +4,471 @@
+  are mapped to ``0`` and ``1``.
+- **optimization:** Booleans are clearly more efficient than integers and we would like to preserve that extreme
+  advantage.
+- **library support:** Pandas for example makes a difference depending if a column contains boolean or integer data:
+
+  >>> import pandas as pd
+  >>> df = pd.DataFrame({
+  ...     "b": [False, True],
+  ...     "i": [0, 1],
+  ... })
+  >>> df.dtypes
+  b     bool
+  i    int64
+  dtype: object
+  >>> ~df["b"]
+  0     True
+  1    False
+  Name: b, dtype: bool
+  >>> ~df["i"]
+  0   -1
+  1   -2
+  Name: i, dtype: int64
+
+
+Null
+----
+While ``null`` has a semantic meaning, they can easily occur in production due to the type inference that pyarrow has to
+do when working with pandas dataframes:
+
+>>> import pandas as p
+>>> import pyarrow as pa
+>>> df = pd.DataFrame({
+...     "single_value": [None, "foo", None],
+...     "no_value": [None, None, None],
+... })
+>>> schema = pa.Schema.from_pandas(df)
+>>> schema.field_by_name("single_value").type
+DataType(string)
+>>> schema.field_by_name("no_value").type
+DataType(null)
+
+The reason is that string and also data objects are stored as ``object`` columns in pandas, which can contain arbitrary
+python objects. ``None`` acts as a placeholder "missing value". Arrow requires that values in a columns have one single
+type and therefore needs to guess what an ``object`` column should represent (i.e. type inference). If pyarrow does not
+find any non-Null object, it treats the column as ``null``. Sadly, this might be wrong. It could easily also have meant
+to be a ``string`` or ``date32`` column, but pyarrow cannot know that.
+
+To keep things pragmatic, we ignore ``null`` during type checks.
+
+Dictionary Encoding
+-------------------
+Dictionary encoded data is normally produced by Pandas categoricals:
+
+>>> import pandas as pd
+>>> import pyarrow as pa
+>>> df = pd.DataFrame({
+...     "s": pd.Series(["foo", "foo", "bar"]).astype("category"),
+... })
+>>> schema = pa.Schema.from_pandas(df)
+>>> schema.field_by_name("s").type
+DictionaryType(dictionary<values=string, indices=int8, ordered=0>)
+
+They have the form ``dictionary[T, I, O]`` where ``T`` represents the value type, ``I`` the index type (mostly integers)
+and ``O`` flags if the index is ordered or not.
+
+Since categoricals are, in our opinion, a pure optimization and do not alter the nature of the data, we treat
+dictionary-encoded data like the values they encode. So ``dictionary[T1, I1, O1]`` is compatible with ``T2`` if ``T1``
+and ``T2`` are compatible. This also means that it is compatible with ``dictionary[T2, I2, O2]``. Note that the ordered
+flag and the index data type are ignored. So the values in the example shown above are treated like ``string``.
+
+Normalization
+-------------
+Following the outlined guidelines, we can write down the following normalization rule set:
```

Comment:
Not sure if here is the best place but is a link to the arrow<->pandas type mapping helpful here? https://arrow.apache.org/docs/python/pandas.html#type-differences

Reply:
added as note to the "base type system" section